### PR TITLE
Add warning message for UNION queries without ALL/DISTINCT keyword

### PR DESCRIPTION
**Pull request by prithvip**
Co-authored-by: Sundeep Katepalli

```
== RELEASE NOTES ==

General Changes
* Added new warning message for UNION queries without ALL/DISTINCT keyword.
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -1254,10 +1255,24 @@ private boolean isExpensiveUnionDistinct(SetOperation setOperation, Type[] outpu
                                             type instanceof RowType);
         }
 
+        @Override
+        protected Scope visitUnion(Union node, Optional<Scope> scope)
+        {
+            if (!node.isDistinct().isPresent()) {
+                warningCollector.add(new PrestoWarning(PARSER_WARNING,
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1262]: @kaikalur Should this be a parser warning or a performance warning? I'm debating on this...

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -1254,10 +1255,24 @@ private boolean isExpensiveUnionDistinct(SetOperation setOperation, Type[] outpu
...
+                        "UNION specified without ALL or DISTINCT" +
+                        " keyword is equivalent to UNION DISTINCT, which is computationally expensive. " +
+                        "Consider using UNION ALL when possible, or specifically add keyword DISTINCT to stop " +
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1265]: Let's not give the option to add DISTINCT here. So encourage them to add ALL. Most people would just do it. And if it's really needed they will figure it out or ask us. So reword it as something like:

"UNION without the ALL keyword is a computationally expensive operation and most often unnecessarry. Please add the ALL keyword."

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 1 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1262]: Should be a performance warning given by the parser :)

**Comment by rongrong** [comment]
> This is a bad idea.. it will complain about queries with small tables that are not expensive

Adding the keyword `DISTINCT` would make the warning go away. And even for small tables, using `UNION DISTINCT` when not necessary still adds small overhead.

**Comment by rongrong** [comment]
> Actually, I think you can do it more succinctly and elegantly in the ASTBuilder. Reduces the size of the PR :) So look into that.

AstBuilder does not have access to `WarningCollector`.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1265]: That's too much. Do you really want the community to open issues here to ask how to make this warning go away?

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1265]: Let's change it to "Consider using UNION ALL when possible, or specifically add the keyword DISTINCT if absolutely necessary".

**Comment by kaikalur** [comment]
> > Actually, I think you can do it more succinctly and elegantly in the ASTBuilder. Reduces the size of the PR :) So look into that.
> 
> AstBuilder does not have access to `WarningCollector`.

:(

**Comment by iceted** [comment]
> Adding the keyword DISTINCT would make the warning go away. And even for small tables, using UNION DISTINCT when not necessary still adds small overhead.

not very user friendly.. every other database works fine with UNION and people are used to it. 

how will I turn it off without asking all my users to modify their queries?

**Comment by rongrong** [comment]
> not very user friendly.. every other database works fine with UNION and people are used to it.
> 
> how will I turn it off without asking all my users to modify their queries?

There's a session property `warning_handling`. You can do `set session warning_handling = 'SUPPRESS'`, or overwrite the default with config property `warning-collector.warning-handling`.

**Comment by iceted** [comment]
i don't want to turn off all warnings

**Comment by rongrong** [comment]
> i don't want to turn off all warnings

What about extending the options for `warning_handling` to include `SUPPRESS` (suppress all), and `SUPPRESS_PERFORMANCE_WARNING`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 1 above)
Comment (prithvip) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Changed to performance warning

**Comment by prithvip** [comment]
Thanks for the review. I've changed the warning message and changed the warning type from parser warning to performance warning.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (prithvip) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: > Let's change it to "Consider using UNION ALL when possible, or specifically add the keyword DISTINCT if absolutely necessary".

Done.

**Comment by iceted** [comment]
why did this get merged when there are still concerns? please revert

**Comment by rongrong** [comment]
> why did this get merged when there are still concerns? please revert

Hi @iceted We proposed alternative solutions 1 week ago but you didn't reply so we moved forward. We've seen a lot of inefficiencies due to this issue so we think it's reasonable to have a performance warning. A warning would not affect the run time in any way other than surfacing a message unless you choose to treat it as an error. So this warning would do nothing more than showing to your users that message. **Query would run as before.** If you have a proposal on how to improve the experience for your users we are happy to hear.

**Pull request merged**